### ISSUE TITLE
PubSub Subscription 'ackDeadline' and/or 'ackDeadlineSeconds'

### ISSUE THREAD

**Issue by ChrisBartonLC** [GoogleCloudPlatform/google-cloud-node#2567]
I am receiving "Error: .google.pubsub.v1.Subscription#ackDeadline is not a field: undefined" when using ackDeadline but switching the options to 'ackDeadlineSeconds' stops the error and reports the correct value in the GCP web console but seems to set the deadline to a very small value because it does not seem to have any effect i.e. I can make repeated calls and receive the same messages.

          let topic = client.pubsub.topic(topic_name)
          let opts = {
            ackDeadline: 60000,
            flowControl: { maxMessages: 5 },
          }
          topic.createSubscription('foo', opts)

but switching the options to the following works:

          let opts = {
            ackDeadlineSeconds: 60,
            flowControl: { maxMessages: 5 },
          }

**Comment by callmehiphop** [comment]
Oh no! Sorry about that, that's some old documentation that was meant to be removed. Instead of redelivering messages, we modify the ack deadline for you, so that option shouldn't be used.

**Comment by chris-barton-localcover** [comment]
But the field 'ackDeadline' is also used in the code and hard coded to "this.ackDeadline = 10000;":

https://github.com/GoogleCloudPlatform/google-cloud-node/blob/pubsub-0.14.0/packages/pubsub/src/subscription.js#L446

**Comment by chris-barton-localcover** [comment]
Also the attribute doesn't seem to have an effect, i.e. I can make repeated calls and receive the same messages.

**Comment by callmehiphop** [comment]
That is correct

**Comment by lukesneeringer** [comment]
@ChrisBartonLC: Some more detail on this. The TL;DR version of it is "You can never have to think about ack deadlines and we will do the right thing."

Basically how it works under the hood is: when the client library receives a message, it automatically sets the ack deadline to 10 seconds (as you observed), _and_ we have a process that runs on a cadence that _continues to extend it_ until you `ack` or `nack`. If something bad happens and, say,  your application crashes, then obviously we stop extending (because we crashed) and you get the messages again.

It is not quite as simple as firing off a request every 10 seconds. That is the starting value. We have some logic where we time how long it takes you to ack all of your past messages and make the deadline appropriate to that. So, if you ack a all of your messages after a minute, then we will start extending the deadline for a minute. (We set the deadline to the 99th percentile of ack times since you opened the subscription -- and we keep extending it, so in theory you should never get a redelivery based on a client library failure so long as the client does not crash or something similar.) That way if you have longer-running jobs, you will not be likely to get messages redelivered because of (for example) network latency, but if you have short messages, we minimize how long you wait to get the message again if something goes wrong.

If for whatever reason this is a problem for your use case, then we would love to hear details. Assuming you have a common use case, however, you should be able to just never think about having to do any of that work yourself, and only worry about `ack` and `nack`.

**Comment by lukesneeringer** [comment]
I am also going to close this under the assumption that the issue is resolved, but feel free to reopen (or open a new issue) if my assessment is incorrect for whatever reason. :-)

**Comment by ChrisBartonLC** [comment]
It seems to work fine but FYI 'So, if you ack a all of your messages after a minute' in a Cloud Function this code runs for seconds not minutes. I've seen other "google-cloud-node" clients e.g the Spanner one have issues when running in Cloud Functions because of assumptions about usage which don't apply e.g. a lot of the caching logic just gets in the way.

**Comment by ChrisBartonLC** [comment]
@callmehiphop Thinking about this some more...What would happen if I have 10 messages that couldn't be parsed and for which the ACK deadline would always pass because they caused an exception. Would these messages always be available for consumption after 10 seconds? Would these have the potential to block all message processing if say I only had one worker and that worker was set to process a maximum of 10 messages? I mention this because in the past on SQS I have used delay queues to guard against this and because PubSub doesn't have delay queues I set the ACK deadline to some time >> larger that the rate at which the messages were being processed.

**Comment by callmehiphop** [comment]
@ChrisBartonLC In the scenario you described I would suggest calling `nack()` on any message you did not want to hang around.

**Comment by ChrisBartonLC** [comment]
Thanks

**Comment by Schibum** [comment]
It would be nice if there was a way to set a max deadline for the current behaviour, to get a faster retries when a process dies for long-running jobs (say 99 ack percentile is 30min, but I want to get retries after 1min).
Or what do you think about making the auto ack deadline extension optional and exposing modifyAckDeadline again? 
The behaviour I want to achieve is a quicker retry of typically long-running jobs with exponential back-off for some errors. With the current implementation, re-delivery seems to be fixed at 0.9*99 percentile of past ack-duration distribution.

**Comment by callmehiphop** [comment]
@Schibum I think allowing users to opt out of the auto-modifying is a great idea, would you be interested in putting together a PR?

**Comment by ChrisBartonLC** [comment]
I would really like to be able to opt out of the auto ack deadline extension, I find that when we get a rogue message that we nack it still gets re-delivered several times before the nack takes effect.

**Comment by amoretspero** [comment]
I have few question about automatic ack deadline extending.

First, is ack deadline is different between per-subscription and per-message? 

Second,
Say we have a root job that spawns many child jobs (more than a thousand) and even that child jobs spawn number of grand-children jobs (10-100 jobs for each of child jobs). We may assume that any job of this kind would take some time, more than default ack deadline time (10secs). But we are sure that this job ends within 10 mins. 

For this root job, I think ack deadline will be extended automatically while job is processed. But at server, as the root job does not acked nor nacked, ack deadline would be 10 secs. Then is it possible the root job is re-received at any subscriber? 

Correction for any misunderstandings are appreciated.

**Comment by callmehiphop** [comment]
> First, is ack deadline is different between per-subscription and per-message?

Each Subscription object collects metrics on the how long it took to ack a message and uses those metrics to guess an average ack time, this ack time is then applied to all messages received on that specific Subscription object and they are not shared with other Subscriptions.

> Then is it possible the root job is re-received at any subscriber?

Once a message is delivered to a Subscription object I believe it will not be re-delivered (anywhere) until the deadline has been hit. Since we automatically modify the deadline for you, the message should not be redelivered unless you close your Subscription object.

Hopefully that answers your question @amoretspero!

**Comment by lfallet** [comment]
Thanks @lukesneeringer @callmehiphop for these ackDeadline details, but I'm confused and don't understand how it integrates with the whole PubSub system:
- I'm surprised there is no mention of this automatic deadline handling in the official docs
- is this 99percentile of ackDeadline on NodeJS client only?
- for time consuming message handling (say 5 minutes), is it possible to set the deadline before starting, in order to prevent a warm up? (first message might be resent many times before the ackDeadline is properly set)

**Comment by callmehiphop** [comment]
> I'm surprised there is no mention of this automatic deadline handling in the official docs

There is mention, but it often gets overlooked. I'm currently working at improving our documentation for this feature at the moment.

> is this 99percentile of ackDeadline on NodeJS client only?

I'm pretty sure most of them do, the Python client does for sure. This was a request made on behalf of the PubSub API team.

> is it possible to set the deadline before starting

Currently it is not, but there is a PR out (https://github.com/googleapis/nodejs-pubsub/pull/388) with a multitude of changes, one of which will allow `ackDeadline` to be set.
